pkg/graveler/graveler.go

```diff
@@ -1879,7 +1880,116 @@ func (g *KVGraveler) Revert(ctx context.Context, repositoryID RepositoryID, bran
 }
 
 func (g *KVGraveler) Merge(ctx context.Context, repositoryID RepositoryID, destination BranchID, source Ref, commitParams CommitParams, strategy string) (CommitID, error) {
-	return g.db.Merge(ctx, repositoryID, destination, source, commitParams, strategy)
+	var (
+		preRunID         string
+		storageNamespace StorageNamespace
+		commit           Commit
+		commitID         CommitID
+	)
+
+	repo, err := g.RefManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return "", err
+	}
+	storageNamespace = repo.StorageNamespace
+
+	// No retries on any failure during the merge. If the branch changed, it's either that commit is in progress, commit occurred,
+	// or some other branch changing operation. If commit is in-progress, then staging area wasn't empty after we checked so not retrying is ok.
+	// If another commit/merge succeeded, then the user should decide whether to retry the merge.
+	err = g.RefManager.BranchUpdate(ctx, repositoryID, destination, func(branch *Branch) (*Branch, error) {
+		empty, err := g.stagingEmpty(ctx, repositoryID, repo, branch)
+		if err != nil {
+			return nil, fmt.Errorf("check if staging empty: %w", err)
+		}
+		if !empty {
+			return nil, fmt.Errorf("%s: %w", destination, ErrDirtyBranch)
+		}
+		fromCommit, toCommit, baseCommit, err := g.getCommitsForMerge(ctx, repositoryID, source, Ref(destination))
+		if err != nil {
+			return nil, err
+		}
+		g.log.WithFields(logging.Fields{
+			"repository":             source,
+			"source":                 source,
+			"destination":            destination,
+			"source_meta_range":      fromCommit.MetaRangeID,
+			"destination_meta_range": toCommit.MetaRangeID,
+			"base_meta_range":        baseCommit.MetaRangeID,
+		}).Trace("Merge")
+		mergeStrategy := MergeStrategyNone
+		if strategy == MergeStrategyDestWins {
+			mergeStrategy = MergeStrategyDest
+		}
+		if strategy == MergeStrategySrcWins {
+			mergeStrategy = MergeStrategySource
+		}
+		metaRangeID, err := g.CommittedManager.Merge(ctx, storageNamespace, toCommit.MetaRangeID, fromCommit.MetaRangeID, baseCommit.MetaRangeID, mergeStrategy)
+		if err != nil {
+			if !errors.Is(err, ErrUserVisible) {
```

Comment:
Not sure what the original purpose was - but this Err is checked but never raised. Maybe @arielshaqed can elaborate more about this error

Reply:
graveler/errors.go wraps numerous user-visible errors around this error, and AFAIK we _do_ return those errors.  These are errors that get pushed to the user, so must not be further wrapped with debug info.  Part of fixing an old bug where we would drop an inordinate amount of information, some of it potentially sensitive, into errors reported to the user.

Reply:
Thanks - I missed the wrapping - BTW is the TODO in errors.go still relevant?